src/Server.php

```diff
@@ -1,95 +1,175 @@
+    private $port = 53;
+    private $ttl = 300;
+    private $maxPacketLength = 512;
 
     /**
-     * @var AbstractStorageProvider
+     * @var ResolverInterface
      */
-    private $ds_storage;
+    private $resolver;
+
+    protected $eventSubscribes = [];
 
-    public function __construct($ds_storage, $bind_ip = '0.0.0.0', $bind_port = 53, $default_ttl = 300, $max_packet_len = 512)
+    /**
+     * Server constructor.
+     *
+     * @param ResolverInterface $resolver
+     * @param array             $config
+     */
+    public function __construct(ResolverInterface $resolver, array $config = [])
     {
-        $this->DS_PORT = $bind_port;
-        $this->DS_IP = $bind_ip;
-        $this->DS_TTL = $default_ttl;
-        $this->DS_MAX_LENGTH = $max_packet_len;
-        $this->ds_storage = $ds_storage;
+        // @todo validate configuration validate ip, port, max packet and ttl as positive integers
+
+        if (isset($config['server']) && $server = $config['server']) {
+            $this->ip = isset($server['bind']) ? $server['bind'] : $this->ip;
+            $this->port = isset($server['port']) ? $server['port'] : $this->port;
+            $this->ttl = isset($server['ttl']) ? $server['ttl'] : $this->ttl;
+            $this->maxPacketLength = isset($server['max_packet_length']) ? $server['max_packet_length'] : $this->maxPacketLength;
+        }
+
+        $this->resolver = $resolver;
 
         ini_set('display_errors', true);
         ini_set('error_reporting', E_ALL);
 
-        set_error_handler(array($this, 'ds_error'), E_ALL);
+        set_error_handler([$this, 'errorHandler'], E_ALL);
         set_time_limit(0);
 
         if (!extension_loaded('sockets') || !function_exists('socket_create')) {
-            $this->ds_error(E_USER_ERROR, 'Socket extension or function not found.', __FILE__, __LINE__);
+            $this->errorHandler(E_USER_ERROR, 'Socket extension or function not found.', __FILE__, __LINE__);
         }
     }
 
-    public function start()
+    /**
+     * Register event subscriber to class.
+     *
+     * @param EventSubscriberInterface $subscriber
+     */
+    public function registerEventSubscriber(EventSubscriberInterface $subscriber)
     {
-        $ds_socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+        $this->eventSubscribes[] = $subscriber;
+    }
 
-        if (!$ds_socket) {
-            $error = sprintf('Cannot create socket (socket error: %s).', socket_strerror(socket_last_error($ds_socket)));
-            $this->ds_error(E_USER_ERROR, $error, __FILE__, __LINE__);
+    /**
+     * Starts DNS server.
+     */
+    public function run()
+    {
+        $socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+
+        if (!$socket) {
+            $error = sprintf(
+                'Cannot create socket (socket error: %s).',
+                socket_strerror(socket_last_error($socket))
+            );
+            $this->errorHandler(E_USER_ERROR, $error, __FILE__, __LINE__);
```

Comment:
Same as above, should probably return from run() when this error happens.

Reply:
errorHandler function had die at the end, I didn't liked it so I removed it. Maybe most correct way would be to throw an Exception?

Reply:
Yep, I like the Exception idea.